File: algorithms/hard/paint_house_iii.ex
defmodule Solution do
  @max 10_000_000

  @spec min_cost(
          houses :: [integer],
          cost :: [[integer]],
          m :: integer,
          n :: integer,
          target :: integer
        ) :: integer
  def min_cost(houses, cost, m, n, target) do
    hs =
      houses
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {h, i}, acc ->
        Map.put(acc, i, h)
      end)

    cs =
      cost
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {row, i}, acc ->
        row
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {c, j}, acc ->
          Map.put(acc, {i, j}, c)
        end)
      end)

    dp =
      for i <- 0..m, j <- 0..m, k <- 0..n, reduce: %{} do
        acc ->
          if i == 0 && j == 0 do
            Map.put(acc, {i, j, k}, 0)
          else
            Map.put(acc, {i, j, k}, @max)
          end
      end

    dp =
      for i <- 1..m, j <- 1..i, reduce: dp do
        acc ->
          if hs[i - 1] == 0 do
            for color <- 1..n, reduce: acc do
              acc ->
                cal_cost(acc, n, cs[{i - 1, color - 1}], i, j, color)
            end
          else
            cal_cost(acc, n, 0, i, j, hs[i - 1])
          end
      end

    for k <- 1..n, reduce: @max do
      acc -> min(acc, dp[{m, target, k}])
    end
    |> then(fn ans ->
      if ans == @max, do: -1, else: ans
    end)
  end

  def cal_cost(dp, n, cost, house, group, color) do
    for pre_color <- 1..n, reduce: dp[{house, group, color}] do
      min_val ->
        pre_cost =
          if pre_color == color do
            dp[{house - 1, group, pre_color}]
          else
            dp[{house - 1, group - 1, pre_color}]
          end

        min(min_val, pre_cost + cost)
    end
    |> then(&Map.put(dp, {house, group, color}, &1))
  end
end

File: algorithms/hard/paint_house_iii.rs
fn calc_cost(dp: &mut Vec<Vec<Vec<i32>>>, n: usize, cost: i32, h: usize, g: usize, c: usize) {
    for pre_color in 1..=n {
        let pre_cost = if pre_color == c {
            dp[h - 1][g][pre_color]
        } else {
            dp[h - 1][g - 1][pre_color]
        };

        dp[h][g][c] = dp[h][g][c].min(pre_cost + cost);
    }
}

impl Solution {
    pub fn min_cost(houses: Vec<i32>, cost: Vec<Vec<i32>>, m: i32, n: i32, target: i32) -> i32 {
        let m = m as usize;
        let n = n as usize;
        let max = 10000000;
        let mut dp = vec![vec![vec![max; n + 1]; m + 1]; m + 1];
        for c in 1..=n {
            dp[0][0][c] = 0;
        }

        for h in 1..=m {
            for g in 1..=h {
                if houses[h - 1] == 0 {
                    for color in 1..=n {
                        calc_cost(&mut dp, n, cost[h - 1][color - 1], h, g, color);
                    }
                } else {
                    calc_cost(&mut dp, n, 0, h, g, houses[h - 1] as usize);
                }
            }
        }

        let mut ans = max;
        for c in 1..=n {
            ans = ans.min(dp[m][target as usize][c]);
        }

        if ans == max {
            -1
        } else {
            ans
        }
    }
}
